File: languages/zig/fatigue_calculator.zig
// fatigue_calculator.zig
// Zig implementation of fatigue calculation algorithms
// Zero-cost abstractions with safety guarantees

const std = @import("std");
const Allocator = std.mem.Allocator;

pub const ActivityLevel = enum {
    sedentary,
    active,
    high_intensity,
};

pub const FatigueLevel = enum {
    mild,
    moderate,
    severe,
};

pub const FatigueError = error{
    NegativeScore,
    InvalidLevel,
    EmptySlice,
    AllocationFailed,
    MathError,
};

pub const FatigueProfile = struct {
    sedentary_limit: f64 = 5.0,
    active_limit: f64 = 8.0,
    high_intensity_limit: f64 = 10.0,
    
    fn getLimit(self: FatigueProfile, level: ActivityLevel) f64 {
        return switch (level) {
            .sedentary => self.sedentary_limit,
            .active => self.active_limit,
            .high_intensity => self.high_intensity_limit,
        };
    }
};

pub const FatigueSummary = struct {
    total: f64,
    count: usize,
    average: f64,
    maximum: f64,
    minimum: f64,
    recovered: bool,
};

pub const FatigueCalculator = struct {
    profile: FatigueProfile,
    allocator: Allocator,
    
    pub fn init(allocator: Allocator, profile: FatigueProfile) FatigueCalculator {
        return FatigueCalculator{
            .profile = profile,
            .allocator = allocator,
        };
    }
    
    pub fn deinit(self: *FatigueCalculator) void {
        _ = self; // Nothing to free currently
    }
    
    pub fn calculateFatigue(self: FatigueCalculator, score: f64, level: ActivityLevel) FatigueError!f64 {
        if (score < 0) {
            return FatigueError.NegativeScore;
        }
        const limit = self.profile.getLimit(level);
        return if (score < limit) score else limit;
    }
    
    pub fn calculateRecovery(self: FatigueCalculator, fatigue: f64, sleep: f64) FatigueError!f64 {
        if (fatigue < 0 or sleep < 0) {
            return FatigueError.NegativeScore;
        }
        const recovered = fatigue - (sleep * 0.1);
        return if (recovered > 0) recovered else 0;
    }
    
    pub fn summarize(self: FatigueCalculator, scores: []const f64) FatigueError!FatigueSummary {
        if (scores.len == 0) {
            return FatigueError.EmptySlice;
        }
        
        var total: f64 = 0;
        var maximum: f64 = scores[0];
        var minimum: f64 = scores[0];
        
        for (scores) |score| {
            total += score;
            if (score > maximum) maximum = score;
            if (score < minimum) minimum = score;
        }
        
        const average = total / @as(f64, @floatFromInt(scores.len));
        const recovered = average < 5.0;
        
        return FatigueSummary{
            .total = total,
            .count = scores.len,
            .average = average,
            .maximum = maximum,
            .minimum = minimum,
            .recovered = recovered,
        };
    }
    
    pub fn computeIntensityMetrics(_: FatigueCalculator, base: f64, multiplier: f64) FatigueError!f64 {
        if (multiplier < 0) {
            return FatigueError.InvalidLevel;
        }
        return base * multiplier;
    }
    
    pub fn adjustForActivityLevel(self: FatigueCalculator, base: f64, activity: ActivityLevel) FatigueError!f64 {
        const factors = .{ .sedentary = 1.0, .active = 1.2, .high_intensity = 1.5 };
        const factor = switch (activity) {
            .sedentary => factors.sedentary,
            .active => factors.active,
            .high_intensity => factors.high_intensity,
        };
        const limit = self.profile.getLimit(activity);
        const adjusted = base * factor;
        return if (adjusted < limit) adjusted else limit;
    }
    
    pub fn verifyThreshold(_: FatigueCalculator, computed: f64, threshold: f64) FatigueError!bool {
        if (computed < 0 or threshold < 0) {
            return FatigueError.NegativeScore;
        }
        return computed >= threshold;
    }
    
    pub fn metricsToJSON(self: *FatigueCalculator, metrics: FatigueSummary) FatigueError![]u8 {
        var json = std.ArrayList(u8).init(self.allocator);
        defer json.deinit();
        
        // Manual JSON serialization for Zig
        const writer = json.writer();
        
        writer.print("{{", .{}) catch return FatigueError.AllocationFailed;
        writer.print("\"total\":{d:.2},", .{metrics.total}) catch return FatigueError.AllocationFailed;
        writer.print("\"count\":{d},", .{metrics.count}) catch return FatigueError.AllocationFailed;
        writer.print("\"average\":{d:.2},", .{metrics.average}) catch return FatigueError.AllocationFailed;
        writer.print("\"maximum\":{d:.2},", .{metrics.maximum}) catch return FatigueError.AllocationFailed;
        writer.print("\"minimum\":{d:.2},", .{metrics.minimum}) catch return FatigueError.AllocationFailed;
        writer.print("\"recovered\":{s}", .{if (metrics.recovered) "true" else "false"}) catch return FatigueError.AllocationFailed;
        writer.print("}}", .{}) catch return FatigueError.AllocationFailed;
        
        return json.toOwnedSlice() catch FatigueError.AllocationFailed;
    }
    
    pub fn metricsFromJSON(_: FatigueCalculator, json_str: []const u8) FatigueError!FatigueSummary {
        // Simplified JSON parsing - assumes valid JSON with known structure
        // In production would use proper JSON parser
        _ = json_str;
        return FatigueError.InvalidLevel; // Placeholder for complex parsing
    }
    
    pub fn serializeMetrics(self: *FatigueCalculator, metrics: FatigueSummary) FatigueError![]u8 {
        return self.metricsToJSON(metrics);
    }
    
    pub fn categorizeFatigueLevel(_: FatigueCalculator, score: f64) FatigueError!FatigueLevel {
        if (score < 0) {
            return FatigueError.NegativeScore;
        }
        return if (score < 3.0) FatigueLevel.mild
            else if (score < 7.0) FatigueLevel.moderate
            else FatigueLevel.severe;
    }
    
    pub fn computeZScore(_: FatigueCalculator, values: []const f64) FatigueError!f64 {
        if (values.len == 0) {
            return FatigueError.EmptySlice;
        }
        
        var sum: f64 = 0;
        for (values) |v| {
            sum += v;
        }
        const mean = sum / @as(f64, @floatFromInt(values.len));
        
        var variance_sum: f64 = 0;
        for (values) |v| {
            const diff = v - mean;
            variance_sum += diff * diff;
        }
        const variance = variance_sum / @as(f64, @floatFromInt(values.len));
        const std_dev = std.math.sqrt(variance);
        
        return if (std_dev == 0) 0 else (values[0] - mean) / std_dev;
    }
    
    pub fn calculateRMSE(_: FatigueCalculator, predicted: []const f64, actual: []const f64) FatigueError!f64 {
        if (predicted.len != actual.len or predicted.len == 0) {
            return FatigueError.EmptySlice;
        }
        
        var sum_squared_error: f64 = 0;
        var i: usize = 0;
        while (i < predicted.len) : (i += 1) {
            const err = predicted[i] - actual[i];
            sum_squared_error += err * err;
        }
        
        const mse = sum_squared_error / @as(f64, @floatFromInt(predicted.len));
        return std.math.sqrt(mse);
    }
    
    pub fn generateForecastTrend(self: *FatigueCalculator, historical: []const f64, periods: usize) FatigueError![]f64 {
        if (historical.len == 0 or periods == 0) {
            return FatigueError.EmptySlice;
        }
        
        const forecasts = self.allocator.alloc(f64, periods) catch {
            return FatigueError.AllocationFailed;
        };
        
        const last = historical[historical.len - 1];
        
        var i: usize = 0;
        while (i < periods) : (i += 1) {
            forecasts[i] = last + @as(f64, @floatFromInt(i + 1));
        }
        
        return forecasts;
    }
};

// Export for package use
pub fn createCalculator(allocator: Allocator, profile: FatigueProfile) FatigueCalculator {
    return FatigueCalculator.init(allocator, profile);
}

File: languages/lua/FatigueCalculator.lua
-- FatigueCalculator.lua
-- Lua implementation of fatigue calculation algorithms
-- Portable across Lua 5.1-5.4, LuaJIT, and embedded contexts

local FatigueCalculator = {}
FatigueCalculator.__index = FatigueCalculator

-- Activity levels
FatigueCalculator.ActivityLevel = {
    SEDENTARY = "sedentary",
    ACTIVE = "active",
    HIGH_INTENSITY = "high_intensity"
}

-- Fatigue levels
FatigueCalculator.FatigueLevel = {
    MILD = "mild",
    MODERATE = "moderate",
    SEVERE = "severe"
}

-- Error types
FatigueCalculator.Errors = {
    NEGATIVE_SCORE = "negative score",
    INVALID_LEVEL = "invalid level",
    EMPTY_SLICE = "empty slice"
}

-- Default profile
local DEFAULT_PROFILE = {
    sedentary_limit = 5.0,
    active_limit = 8.0,
    high_intensity_limit = 10.0
}

-- Constructor
---@param profile table|nil Custom fatigue profile or nil for defaults
---@return FatigueCalculator
function FatigueCalculator.new(profile)
    local self = setmetatable({}, FatigueCalculator)
    self.profile = profile or DEFAULT_PROFILE
    return self
end

-- Get limit for activity level
---@param level string Activity level
---@return number
function FatigueCalculator:get_limit(level)
    if level == self.ActivityLevel.SEDENTARY then
        return self.profile.sedentary_limit
    elseif level == self.ActivityLevel.ACTIVE then
        return self.profile.active_limit
    elseif level == self.ActivityLevel.HIGH_INTENSITY then
        return self.profile.high_intensity_limit
    else
        error(self.Errors.INVALID_LEVEL)
    end
end

-- Calculate fatigue score
---@param score number Input score
---@param level string Activity level
---@return table Result with value or error
function FatigueCalculator:calculate_fatigue(score, level)
    if score < 0 then
        return { success = false, error = self.Errors.NEGATIVE_SCORE }
    end
    
    local limit = self:get_limit(level)
    return { success = true, value = math.min(score, limit) }
end

-- Calculate recovery
---@param fatigue number Current fatigue
---@param sleep number Sleep hours
---@return table Result with recovered fatigue or error
function FatigueCalculator:calculate_recovery(fatigue, sleep)
    if fatigue < 0 or sleep < 0 then
        return { success = false, error = self.Errors.NEGATIVE_SCORE }
    end
    
    local recovered = math.max(fatigue - (sleep * 0.1), 0)
    return { success = true, value = recovered }
end

-- Summarize scores
---@param scores table Array of scores
---@return table Result with summary or error
function FatigueCalculator:summarize(scores)
    if #scores == 0 then
        return { success = false, error = self.Errors.EMPTY_SLICE }
    end
    
    local total = 0
    local maximum = scores[1]
    local minimum = scores[1]
    
    for _, score in ipairs(scores) do
        total = total + score
        if score > maximum then maximum = score end
        if score < minimum then minimum = score end
    end
    
    local average = total / #scores
    
    return {
        success = true,
        value = {
            total = total,
            count = #scores,
            average = average,
            maximum = maximum,
            minimum = minimum,
            recovered = average < 5.0
        }
    }
end

-- Compute intensity metrics
---@param base number Base value
---@param multiplier number Multiplier
---@return table Result with intensity or error
function FatigueCalculator:compute_intensity_metrics(base, multiplier)
    if multiplier < 0 then
        return { success = false, error = self.Errors.INVALID_LEVEL }
    end
    return { success = true, value = base * multiplier }
end

-- Adjust for activity level
---@param base number Base value
---@param activity string Activity level
---@return table Result with adjusted value or error
function FatigueCalculator:adjust_for_activity_level(base, activity)
    local factors = {
        [self.ActivityLevel.SEDENTARY] = 1.0,
        [self.ActivityLevel.ACTIVE] = 1.2,
        [self.ActivityLevel.HIGH_INTENSITY] = 1.5
    }
    
    local factor = factors[activity]
    if not factor then
        return { success = false, error = self.Errors.INVALID_LEVEL }
    end
    
    return { success = true, value = base * factor }
end

-- Verify threshold
---@param computed number Computed value
---@param threshold number Threshold to check against
---@return table Result with boolean or error
function FatigueCalculator:verify_threshold(computed, threshold)
    if computed < 0 or threshold < 0 then
        return { success = false, error = self.Errors.NEGATIVE_SCORE }
    end
    return { success = true, value = computed >= threshold }
end

-- Convert metrics to JSON
---@param metrics table Metrics summary
---@return string JSON string
function FatigueCalculator:metrics_to_json(metrics)
    local parts = {}
    table.insert(parts, string.format('"total":%.2f', metrics.total))
    table.insert(parts, string.format('"count":%d', metrics.count))
    table.insert(parts, string.format('"average":%.2f', metrics.average))
    table.insert(parts, string.format('"maximum":%.2f', metrics.maximum))
    table.insert(parts, string.format('"minimum":%.2f', metrics.minimum))
    table.insert(parts, string.format('"recovered":%s', tostring(metrics.recovered)))
    return "{" .. table.concat(parts, ",") .. "}"
end

-- Parse metrics from JSON (simplified)
---@param json_str string JSON string
---@return table Result with metrics or error
function FatigueCalculator:metrics_from_json(json_str)
    -- This is a simplified implementation
    -- In production, use a proper JSON library like cjson or dkjson
    local success, metrics = pcall(function()
        -- Very basic parsing - assumes well-formed JSON
        local result = {}
        for k, v in json_str:gmatch('"(%w+)":([%d%.]+)') do
            result[k] = tonumber(v)
        end
        result.recovered = json_str:find('"recovered":true') ~= nil
        return result
    end)
    
    if not success then
        return { success = false, error = "JSON parse error" }
    end
    
    -- Ensure required fields
    if not (metrics.total and metrics.count and metrics.average) then
        return { success = false, error = "Missing required fields" }
    end
    
    return { success = true, value = metrics }
end

-- Serialize metrics
---@param metrics table Metrics to serialize
---@return string JSON string
function FatigueCalculator:serialize_metrics(metrics)
    return self:metrics_to_json(metrics)
end

-- Categorize fatigue level
---@param score number Score to categorize
---@return table Result with level or error
function FatigueCalculator:categorize_fatigue_level(score)
    if score < 0 then
        return { success = false, error = self.Errors.NEGATIVE_SCORE }
    end
    
    if score < 3.0 then
        return { success = true, value = self.FatigueLevel.MILD }
    elseif score < 7.0 then
        return { success = true, value = self.FatigueLevel.MODERATE }
    else
        return { success = true, value = self.FatigueLevel.SEVERE }
    end
end

-- Compute z-score
---@param values table Array of values
---@return table Result with z-score or error
function FatigueCalculator:compute_z_score(values)
    if #values == 0 then
        return { success = false, error = self.Errors.EMPTY_SLICE }
    end
    
    local sum = 0
    for _, v in ipairs(values) do
        sum = sum + v
    end
    local mean = sum / #values
    
    local variance_sum = 0
    for _, v in ipairs(values) do
        local diff = v - mean
        variance_sum = variance_sum + (diff * diff)
    end
    local variance = variance_sum / #values
    local std_dev = math.sqrt(variance)
    
    if std_dev == 0 then
        return { success = true, value = 0 }
    end
    
    return { success = true, value = (values[1] - mean) / std_dev }
end

-- Calculate RMSE
---@param predicted table Predicted values
---@param actual table Actual values
---@return table Result with RMSE or error
function FatigueCalculator:calculate_rmse(predicted, actual)
    if #predicted ~= #actual or #predicted == 0 then
        return { success = false, error = self.Errors.EMPTY_SLICE }
    end
    
    local sum_squared_error = 0
    for i = 1, #predicted do
        local err = predicted[i] - actual[i]
        sum_squared_error = sum_squared_error + (err * err)
    end
    
    local mse = sum_squared_error / #predicted
    return { success = true, value = math.sqrt(mse) }
end

-- Generate forecast trend
---@param historical table Historical data
---@param periods number Number of periods to forecast
---@return table Result with forecasts or error
function FatigueCalculator:generate_forecast_trend(historical, periods)
    if #historical == 0 or periods <= 0 then
        return { success = false, error = self.Errors.EMPTY_SLICE }
    end
    
    local last = historical[#historical]
    local forecasts = {}
    
    for i = 1, periods do
        local forecast = last + i
        table.insert(forecasts, forecast)
    end
    
    return { success = true, value = forecasts }
end

-- Create empty summary
---@return table Empty summary
function FatigueCalculator.empty_summary()
    return {
        total = 0,
        count = 0,
        average = 0,
        maximum = 0,
        minimum = 0,
        recovered = false
    }
end

return FatigueCalculator
